### PR TITLE
PWGPP-570: Merging ionTailMerge branch into master

### PR DESCRIPTION
**Pull request by miranov25**
Please do not approve - I prefer to run a test first 

# Modifications:

* ion tail _ common mode changes related modification in November-March

### After merging with master:
* Debug streaming - down-sampling of cluster  - used for tuning of the cluster error parameterization
https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239113&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239113
  * Ion tail - integral scaling 
https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239114&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239114 
* Write local baseline shift due ion tail and common mode to the clusters:
  * https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239115&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239115
* Multiplicity and dEdx dependent Cluster error parameterization - IMPLEMETED (22.11.2015) see
  
  * https://indico.cern.ch/event/790434/contributions/3647824/attachments/1949646/3235806/PWGPP-570_ErrorParameterizationAndClusterAssciation.pdf 
https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239116&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239116
* Switch to enable dEdx, multiplcity dependent parameterization:
  * https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239129&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239129
* Usage of the missing cluster correction in the dEdx calculation -  IMPLEMETED (22.11.2015) see:
  * https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239117&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239117
  * https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239150&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239150

### CODE REVIEW

**Comment by miranov25** [comment]
+test

**Comment by shahor02** [comment]
+test

**Comment by shahor02** [comment]
@miranov25 I suspect your master was not rebased to remote: I see in the list of your already merged commits by others, like https://github.com/alisw/AliRoot/pull/1070/commits /9a3e83fa069a2ae2a4c770ce788805853063c0fe from Peter

**Comment by fprino** [comment]
Hi Marian, all,

I see modifications in AliAnalysisManager.cxx and AliGenPythiaPlus that are not related to the TPC development.
I think this is related to the 2 latest commits in the list of this PR, namely
"fPythia object not persistent to avoid default constructur be called …" 
"Return kTRUE since kFALSE indicates to Root v6-18-04 real problems"
which are modifications done by Peter and Andreas.
Can you please check?

Ciao, Francesco

**Comment by miranov25** [comment]
Hello Reuben and Francesco. 

Not clear what went wrong while merging in previous attempt.
I m almost sure I update my master before merging
To be sure I started again from clean master.
Could you test again?

Marian
```
[AliPhysics/latest] ~/software/aliceGSI/AliRoot $>  git merge ionTailMerge
Merge made by the 'recursive' strategy.
 STAT/AliTMinuitToolkit.cxx         |  20 +++--
 STAT/AliTMinuitToolkit.h           |   2 +-
 STEER/ESD/AliESDUtils.cxx          |   6 +-
 TPC/TPCbase/AliTPCRecoParam.cxx    |  12 ++-
 TPC/TPCbase/AliTPCRecoParam.h      |  25 +++++-
 TPC/TPCbase/AliTPCcalibDB.cxx      | 132 ++++++++++++++++++------------
 TPC/TPCbase/AliTPCclusterMI.cxx    |  10 ++-
 TPC/TPCbase/AliTPCclusterMI.h      |  11 ++-
 TPC/TPCrec/AliTPCReconstructor.cxx |   1 +
 TPC/TPCrec/AliTPCReconstructor.h   |   5 +-
 TPC/TPCrec/AliTPCseed.cxx          |  13 ++-
 TPC/TPCrec/AliTPCtracker.cxx       | 650 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------------
 TPC/TPCrec/AliTPCtracker.h         |   1 +
 TPC/TPCsim/AliTPCDigitizer.cxx     | 300 ++++++++++++++++++++++++++++++++++++-------------------------------
 test/configuration/configGSI.sh    |   2 +-
 15 files changed, 760 insertions(+), 430 deletions(-)
```

**Comment by chiarazampolli** [comment]
+test

**Comment by miranov25** [comment]
+test

**Comment by shahor02** [comment]
now the extra commits are gone. I suspect the failure is spurious, let's wait for the macos.

**Comment by shahor02** [comment]
@ktf, @MichaelLettrich CI reports a failure for release (immidiately after the push) w/o providing the log, is this this a CI problem?

**Comment by ktf** [comment]
@MichaelLettrich did you restart this worker the other day? I just did it again.

**Comment by ktf** [comment]
(regardless of that, looks like there are conflict when merging...)

**Comment by ktf** [comment]
Checks are now successful, but we cannot merge...

File: TPC/TPCsim/AliTPCDigitizer.cxx [hunk 1]
```diff
@@ -704,149 +698,150 @@ void AliTPCDigitizer::DigitizeWithTailAndCrossTalk(Option_t* option)
...
+      if (graphRes[ires]) {
+        timeResArr->AddAt(graphRes[ires], ires);
+        lastGraph = graphRes[ires];
+      } else {
+        if (lastGraph != nullptr) timeResArr->AddAt(new TGraphErrors(*lastGraph), ires);
+      }
+    }
+    timeResFunc.AddAt(timeResArr, isec); // Fill all trfs into a single TObjArray
+    nIonTailBins = graphRes[0]->GetN();
+    delete trfIndexArr;
```
Comment (shahor02) [inline TPC/TPCsim/AliTPCDigitizer.cxx]: @miranov25 I see here you have overwritten the memory leak I've fixed some time ago.

**Comment by shahor02** [comment]
@ktf thanks, how did you know that there are conflicts (with the github saying there are none)?
In fact I am able to do
``git merge  miranov25/master`` on the up to date master w/o conflicts, but his original ionTailMerge produces lot of conflicts when being rebased on master (including the conflict with my fix for the memory leak).
@miranov25 I gues the best if you close this PR, do
````
git checkout master
git fetch origin && git reset --hard origin/master 
git checkout ionTailMerge
git rebase master
# solve conflicts
git push -f miranov25 master
```` 
then open PR for ionTailMerge

Cheers,
 Ruben

**Comment by ktf** [comment]
I misread the message from Github GUI... It says "Unable to **rebase**". 

I can merge directly without rebasing, if that's ok, but I think in general it is good practice to rebase before merging..

**Comment by shahor02** [comment]
I also think it is better to rebase, right now there is at least 1 memory leak fix reverted.

File: TPC/TPCsim/AliTPCDigitizer.cxx (hunk 1 above)
Comment (miranov25) [reply TPC/TPCsim/AliTPCDigitizer.cxx]: I merged this part in March. I assumed I did it correctly. 
I will rebase again ...

**Comment by miranov25** [comment]
OK.
I will  
```
git checkout master
git fetch origin && git reset --hard origin/master 
git checkout ionTailMerge
git rebase master
# solve conflicts
git push -f miranov25 master
```
and open new pull request

**Comment by miranov25** [comment]
f...
meld is crashing in merging.
f...
```
Normal merge conflict for 'TPC/TPCsim/AliTPCDigitizer.cxx':
  {local}: modified file
  {remote}: modified file
Hit return to start merge resolution tool (meld): bc
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/meld/meldwindow.py", line 309, in on_delete_event
    response = page.on_delete_event()
  File "/usr/lib/python3/dist-packages/meld/filediff.py", line 842, in on_delete_event
    del self._cached_match
AttributeError: _cached_match
```

**Comment by miranov25** [comment]
using clion for resolving conflics  I could merge  - but automatic git merging ignored important modifications from iontailMerge
default merge removed my changes in:
````
      modified:   TPC/TPCbase/AliTPCcalibDB.cxx
        modified:   TPC/TPCrec/AliTPCtracker.cxx
        modified:   TPC/TPCsim/AliTPCDigitizer.cxx
````
one more time ...

**Comment by miranov25** [comment]
# Summarizing:

* git rebase master
  * my modifications in the AliTPCcalibDB and AliTPCtracker are ignored
* git merge master
  * Ruben modifications from AliTPCdigitizer are ignored

**Comment by miranov25** [comment]
We are stuck here - and I will have to leave soon.
My proposed solution use git merge master, and merge Ruben diffs later.
Do you agree?

**Comment by shahor02** [comment]
@miranov25 The memory leak you have reintroduced was relevant for the MC only and was not critical, though there might be others. But the proper solution would be to save your current files where your modifications as lost (AliTPCcalibDB and AliTPCtracker only), rebase to master, then manually fix the diffs between saved and rebased version of files.

If this cannot be done now, for the moment we probably can use merge, but as far as I can see, it will but much more difficult to track down later the fixes lost in the merge: you will need to rebase your branch to current master (f58237542c), resolve the conflicts, edit the diffs between the files where there were conflicts and commit them as new patches.

**Comment by miranov25** [comment]
@shahor02 

I checked both approaches - storing and diffing commits. 
In current situation, git rebase can not be used. All my recent commits, which I described in the JIRA PWGPP-570 are lost (e.g for the TPCrec directory - everything  after 15.10)
````
[AliPhysics/latest] ~/software/aliceGSI/AliRoot/TPC/TPCrec $> wc  *gitlog
  1370   5533  48920 TPCrecMerge.gitlog
  1370   5533  48920 TPCrecOrig.gitlog
  1189   4720  42225 TPCrecrebase.gitlog
````
This time, we have to use merge - and add manual fixes later.

**Comment by shahor02** [comment]
> PWGPP-570 are lost (e.g for the TPCrec directory - everything after 15.10)
I don't understand how this is possible: the last commit in the master: TPC/TPCrec is from 2018, rebase should not override any of your recent commits. Let me try

**Comment by miranov25** [comment]
in case you work on that, I merged AliTPCDigitizer already several times
here it is clear how to solve problems. This part i can provide.

**Comment by shahor02** [comment]
@miranov25 Could you check the branch https://github.com/shahor02/AliRoot/tree/mi_IT, it is your ionTailMerge rebased to master and some conflicts solved. I see all you recent commits in TPCrec.
The branch merges to master w/o problem.
If it is OK, just close this PR and open new one with mi_IT

**Comment by miranov25** [comment]
Merge as you prepared is OK
Closing this request

**Pull request closed without merging**